cirq-google/cirq_google/circuit_serializer.py

```diff
@@ -0,0 +1,657 @@
+            )
+            return msg
+        if isinstance(gate, cirq.CZPowGate):
+            arg_func_langs.float_arg_to_proto(
+                gate.exponent,
+                out=msg.czpowgate.exponent,
+                arg_function_language=arg_function_language,
+            )
+            return msg
+        if isinstance(gate, cirq.ISwapPowGate):
+            arg_func_langs.float_arg_to_proto(
+                gate.exponent,
+                out=msg.iswappowgate.exponent,
+                arg_function_language=arg_function_language,
+            )
+            return msg
+        if isinstance(gate, cirq.FSimGate):
+            arg_func_langs.float_arg_to_proto(
+                gate.theta,
+                out=msg.fsimgate.theta,
+                arg_function_language=arg_function_language,
+            )
+            arg_func_langs.float_arg_to_proto(
+                gate.phi,
+                out=msg.fsimgate.phi,
+                arg_function_language=arg_function_language,
+            )
+            return msg
+        if isinstance(gate, cirq.MeasurementGate):
+            arg_func_langs.arg_to_proto(
+                gate.key,
+                out=msg.measurementgate.key,
+                arg_function_language=arg_function_language,
+            )
+            arg_func_langs.arg_to_proto(
+                gate.invert_mask,
+                out=msg.measurementgate.invert_mask,
+                arg_function_language=arg_function_language,
+            )
+            return msg
+        if isinstance(gate, cirq.WaitGate):
+            arg_func_langs.float_arg_to_proto(
+                gate.duration.total_nanos(),
+                out=msg.waitgate.duration_nanos,
+                arg_function_language=arg_function_language,
+            )
+            return msg
+        raise ValueError(f'Cannot serialize op {op!r} of type {type(gate)}')
+
+    def serialize_circuit_op(
+        self,
+        op: cirq.Operation,
+        msg: Optional[v2.program_pb2.CircuitOperation] = None,
+        *,
+        arg_function_language: Optional[str] = '',
+        constants: Optional[List[v2.program_pb2.Constant]] = None,
+        raw_constants: Optional[Dict[Any, int]] = None,
+    ) -> v2.program_pb2.CircuitOperation:
+        """Serialize a CircuitOperation to cirq.google.api.v2.CircuitOperation proto.
+
+        Args:
+            op: The circuit operation to serialize.
+            msg: An optional proto object to populate with the serialization
+                results.
+            arg_function_language: The `arg_function_language` field from
+                `Program.Language`.
+            constants: The list of previously-serialized Constant protos.
+            raw_constants: A map raw objects to their respective indices in
+                `constants`.
+
+        Returns:
+            The cirq.google.api.v2.CircuitOperation proto.
+        """
+        circuit = getattr(op.untagged, 'circuit', None)
```

Comment:
Should fail if `circuit` is None, so could remove the default here. But I would suggest instead doing an explicit type check:
```python
if not isinstance(op.untagged, cirq.CircuitOperation):
   raise ValueError(...)
circuit = op.untagged.circuit
```
This should also make the cast below unnecessary since we've already checked the type of `op.untagged`.

Reply:
Done.